plugin.rb

```diff
@@ -31,23 +31,58 @@ def self.default_image_url
       "#{Discourse.base_url}/plugins/discourse-patreon-creator/images/patreon-logomark-color-on-white.png"
     end
 
-    def self.store
-      @store ||= PluginStore.new(PLUGIN_NAME)
-    end
-
     def self.get(key)
       store.get(key)
     end
 
+    def self.nsfw_group
+      @nsfw_group = Group.find_by_name(SiteSetting.patreon_creator_nsfw_group)
+    end
+
+    def self.remove(key)
+      store.remove(key)
+    end
+
+    def self.save_user_info(user, email, patreon_id, has_nsfw_campaign)
+      plugin_store_key = "user_info.#{email}"
+      if user
+        user.custom_fields["patreon_id"] = patreon_id
+        user.custom_fields["has_nsfw_campaign"] = has_nsfw_campaign
+        user.save_custom_fields
+
+        # Remove the temp data from the plugin store
+        Patreon.remove(plugin_store_key)
+      else
+        # User has not been created, but they are authorized.
+        # save data in plugin store so we can use in during
+        # This is called during authorization and on the
+        # :user_created callback.
+
+        user_info_record = Patreon.get(plugin_store_key) || {}
+        user_info_record[:has_nsfw_campaign] = has_nsfw_campaign
+        user_info_record[:patreon_id] = patreon_id
+        Patreon.set(plugin_store_key, user_info_record)
+      end
+    end
+
     def self.set(key, value)
       store.set(key, value)
     end
+
+    def self.store
+      @store ||= PluginStore.new(PLUGIN_NAME)
+    end
+
+    def self.add_remove_nsfw_group(user, has_nsfw_campaign)
+      return unless user && Patreon.nsfw_group
+
+      has_nsfw_campaign ? Patreon.nsfw_group.add(user) : Patreon.nsfw_group.remove(user)
```

Comment:
Is adding a user to a group idempotent?

Reply:
@elskwid yes, `Group.add` and `Group.remove` are idempotent.

https://github.com/discourse/discourse/blob/master/app/models/group.rb#L614
https://github.com/discourse/discourse/blob/master/app/models/group.rb#L641

Reply:
Great, thanks.